vault/router.go

```diff
@@ -462,7 +462,12 @@ func (r *Router) MatchingBackend(ctx context.Context, path string) logical.Backe
 	if !ok {
 		return nil
 	}
-	return raw.(*routeEntry).backend
+
+	re := raw.(*routeEntry)
+	re.l.RLock()
+	defer re.l.RUnlock()
+
+	return re.backend
```

Comment:
it might be worthwhile to extract these lines to a helper method so that clients don't have to remember to place a lock when reading `backend`.

Reply:
I am sympathetic to this. However, I have a few concerns

- use of `routeEntry` all happens within the `vault` package so to be sure that clients use the helper methods we would want to refactor this to another package and make the fields private so that access to the fields require getters that hold locks, but
- this code is in the hot path of vault requests so I am hesitant to do too much refactoring without involving the vault Core team

Reply:
I think I will leave this as-is and reach out to Core to see what they think about any more changes to the routeEntry